aiostream/stream/combine.py

```diff
@@ -76,14 +78,22 @@ async def zip(*sources: AsyncIterable[T]) -> AsyncIterator[tuple[T, ...]]:
             await stack.enter_async_context(streamcontext(source)) for source in sources
         ]
         # Loop over items
+        STOP_SENTINEL = object()  # only used in strict mode
         while True:
+            coros = (
+                anext(streamer, STOP_SENTINEL) if strict else anext(streamer)
+                for streamer in streamers
+            )
```

Comment:
This currently fails type checking; and it is indeed a bit difficult to type, considering `anext` can return different things depending on whether `strict` is `True` or not.

Maybe there should just be a higher-level `if strict:` (either just inside the loop or maybe even outside of it) so there are two totally separate code blocks for strict and non-strict mode? :thinking: That might make this whole thing less confusing in general than having them both mixed together.

Reply:
> that might make this whole thing less confusing in general than having them both mixed together.

I fully agree with that, easier to type and easier to read.

Reply:
I've changed it like that now and fixed the typing, for which I still had to resort to a `cast` because AFAICT neither Mypy nor Pyright understand that in a situation like 

```python
ls: list[list[int|str]] = [[1, "a"], [2], ["c"]]

for l in ls:
  if any(isinstance(x, int) for x in l):
    break
  reveal_type(l)
```

the type of `l` can only be `list[str]` for the remainder of the loop iteration because we ruled out `int` :shrug: